contents/why_julia.md

```diff
@@ -69,31 +68,261 @@ Let's proceed then!
+You would need to change your approach with a substancial loss of simplicity (you would probably need to create different functions with different names).
+
+**Now let's do this in Julia**.
+
+In Julia, we don't have classes but we have structures (`struct`) that are meant to be "structured data": they define the kind of information that is embedded in the structure, that is a set of fields (i.e. "properties" or "attributes" in other languages), and then individual instances (or "objects") can be produced each with its own specific values for the fields defined by the structure.
+They differ from the primitive types that are by default defined already inside the core of Julia.
+Because of this, `structs`s are composite types, and usually defined by the user.
+
+First we'll create an `abstract type` named `Researcher`:
+
+```jl
+sc(
+"""
+abstract type Researcher end
+"""
+)
+```
+
+We proceed, similar as before, by creating two derived `struct`s from the `Researcher` abstract type:
+
+```jl
+sc(
+"""
+struct Linguist <: Researcher
+    name::AbstractString
+    age::Int64
+end
+"""
+)
+```
+
+```jl
+sc(
+"""
+struct ComputerScientist <: Researcher
+    name::AbstractString
+    age::Int64
+end
+"""
+)
+```
+
+The final step is to define two new functions that will behave differently depending on which derived `struct` of `Researcher` are the first or second argument:
+
+```jl
+sco(
+"""
+approaches(li::Linguist, cs::ComputerScientist) = "Hey \$(cs.name), wanna do a paper? We need to use APA style."
+
+approaches(cs::ComputerScientist, li::Linguist) = "Hey \$(li.name), wanna do a paper? We need to use IEE style."
+"""
+)
+```
+
+Finally, let's instantiate our two researchers, `noam` and `judea`:
+
+```jl
+sc(
+"""
+noam = Linguist("Noam Chomsky", 92)
+judea = ComputerScientist("Judea Pearl", 84)
+"""
+)
+```
+
+Again, let's see what Noam Chomsky will say when he approaches Judea Pearl with a paper idea:
+
+```jl
+sco(
+"""
+approaches(noam, judea)
+"""
+)
+```
+
+Perfect! It behaves just as we wanted!
+This is **multiple dispatch** and it is an important feature in Julia.
+Multiple dispatch acts on all arguments of a function and defines function behavior based on all argument's types.
+
+Multiple dispatch is a powerful feature that allows us also to extend existing functions or to define custom and complex behavior to new types.
+To show how this works, we'll use another example.
+Suppose you want to define two new `struct`s based on a animals:
+
+```jl
+sc(
+"""
+struct fox end
+struct chicken end
+"""
+)
+```
+
+I want to define addition for both `fox` and `chicken` new types.
+We proceed by defining a new function signature of the `+` from the `Base` module of Julia^[this is just an example for teaching purposes. Don't do this... Rik help me out here, I forgot the name of the thing. It was "type" something, right?]:
```

Comment:
@rikhuijzer I need a little help here. It is an advanced thing that we should keep inside the footnote.

Reply:
Method invalidations you mean?

Reply:
What could possibly also be a nice example is to implement the same syntax for paths as Python allows:
```
julia> import Base: /

julia> function /(a::String, b::String)
            joinpath(a, b)
         end
/ (generic function with 107 methods)

julia> path = homedir() / "b" / "c"
"/home/rik/b/c"
```

Reply:
> Method invalidations you mean?

Yes! Thanks I will add that.



> What could possibly also be a nice example is to implement the same syntax for paths as Python allows:
> 
> ```
> julia> import Base: /
> 
> julia> function /(a::String, b::String)
>             joinpath(a, b)
>          end
> / (generic function with 107 methods)
> 
> julia> path = homedir() / "b" / "c"
> "/home/rik/b/c"
> ```

I have never done this in Python. I mainly use `Path` from `pathlib` (used `os` a long time ago, but soon discovered that it is not a good cross-OS tool). You can add if you want to. The only thing that I would need the examples to be are simple. I've directed the reader to more complex examples in other resources (Storopoli Why Julia one hot vector and Rackaucka's quaternion GPU ODE solver)